[1단계 - 출석 구현] 윌슨(박형균) 미션 제출합니다.

phk1128 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
시간이 부족해서 좋은 설계는 못했어요 구체적인 이유는 다음과 같아요

- 컨트롤러의 책임 분리를 적절히 하지 못해서 코드가 180라인 정도가 됨
- 도메인 및 컨트롤러에서 LocalDate 및 LocalDateTime을 파싱하는 로직이 중복됨
- 라인 수가 10이 넘는 메서드가 4개 존재
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

시간 내주셔서 감사해요! 궁금한점 3가지가 있는데 편하실때 답변 부탁드려요~ 

- 현재 서비스 계층이 없어서 dto 변환을 컨트롤러에서 담당하게 되었는데요. 이 경우 dto 변환을 위한 매퍼 클래스로 리팩토링하는게 좋은 방법일까요??
- 애플리케이션 종료 후 resources 패키지 존재하는 csv파일을 원상태로 복구하는게 좋을까요? 아니면 변경 내용들을 csv파일에 저장하는게 좋을까요 ??
- Attendance 클래스 생성자에서 매개변수로 `yyyy-MM-dd HH:mm` 형태의 문자열을 받아 LocalDateTime으로 파싱하는 작업을 하고 있어요. 여기서 외부에서 날짜를 파싱해서 LocalDateTime을 Attendance 생성자로 넘겨주는게 좋을것 같은데, 베루스의 생각은 어떠신가요??


[프롤로그](https://prolog.techcourse.co.kr/studylogs/3971)

phk1128 committed:
docs(README): 기능 요구 사항 작성 하라

- 파일 입출력 기능 요구 사항 작성

phk1128 committed:
test(FileManagerTest): csv 데이터 입출력 레드 테스트하라

- csv 파일 전체를 읽어온다

phk1128 committed:
test(FileManagerTest): csv 파일 입출력을 그린 테스트하라

- csv 파일 전체를 읽어온다.

phk1128 committed:
feat(FileManager): csv 파일을 읽어와라

- readFileLines() 메서드의 프로덕션 코드를 작성하였다.

phk1128 committed:
docs(README): 완성된 파일 입출력 기능을 체크하라

- 리소스에 있는 csv 파일에 데이터 입력 기능을 완성하였다.

phk1128 committed:
feat(CrewName): 크루 이름 도메인을 구현하라

phk1128 committed:
feat(Attendance): 출석 도메인을 구현하라

- Attendance는 출석을 담당하는 도메인 객체
- AttendanceStatus는 출석의 상태를 나타내는 enum 객체

phk1128 committed:
feat(Crew): 크루 도메인을 구현하라

- ExpulsionStatus는 제적 위험 상태를 나타내는 객체

phk1128 committed:
feat: 크루 출석 기록 확인 기능을 구현하라

phk1128 committed:
feat : 크루의 출석 확인 기능을 구현하라

phk1128 committed:
feat : 날짜에 대한, 새로운 Attendance 수정 기능을 구현하라

phk1128 committed:
feat: 제적 위험자 확인 기능을 구현하라

phk1128 committed:
feat: controller에서 각각의 기능들을 조합하라

phk1128 committed:
fix: InputView에서, 파싱에 관한 예외를 IllegalArgumentException으로 던져라

phk1128 committed:
test: Crew 도메인에 대한 단위 테스트를 작성하라

phk1128 committed:
feat: 이미 출석이 존재한다면, 수정으로 안내하라

changuii committed:
test: AttendanceSystem에 대한 단위 테스트를 작성하라

changuii committed:
fix: 클래스파일을 git에서 제거하라

phk1128 committed:
test(AttendanceTest): Nested로 성공 테스트를 분리하라

phk1128 committed:
refactor: 인덴트가 2인 부분을 1로 수정하라

phk1128 committed:
refactor: 등교일이 아닌 경우 사용자에게 안내 후 초기 화면으로 돌아가라

phk1128 committed:
docs(README): 기능 구현 목록을 체크하라

verus-j commented on src/main/java/domain/AnswerCommand.java:
```diff
@@ -0,0 +1,22 @@
+        this.command = command;
+    }
+
+    public static AnswerCommand findByCommand(final String command) {
+        return Arrays.stream(AnswerCommand.values())
+                .filter(answerCommand -> Objects.equals(answerCommand.command, command))
+                .findAny()
+                .orElseThrow(IllegalArgumentException::new);
```
여기는 테스트가 없네요 ㅠㅠ
느낌상 AnswerCommand는 사용자의 입력을 변환해주는 역할인 거 같은데, 도메인에 있어야하나요??
출석 도메인에 응답을 YES, NO로 변환하는 규칙이 있나요?

phk1128 replied:
이미 출석한 상태에서 출석 확인을 하면 사용자에게 출석 수정으로 이동할 것이냐 묻는 커멘드에요
요구사항에는 나와 있지 않았지만, 사용자 경험 측면에서 좋을것 같아 이처럼 구현했어요
그리고 이 또한 비즈니스 로직에 해당한다고 생각해서 도메인에 두었어요!

verus-j commented on src/main/java/domain/AttendanceSystem.java:
```diff
@@ -0,0 +1,110 @@
+        if (!existTodayAttendanceByCrewName(crewName, LocalDate.of(2024, 12, dayOfMonth))) {
+            throw new IllegalArgumentException("유효하지 않은 날짜입니다.");
+        }
+    }
```
두개 메서드 모두 정상적인 케이스에 대한 테스트가 없네요 ㅠㅠ

verus-j commented on src/main/java/domain/AttendanceSystem.java:
```diff
@@ -0,0 +1,110 @@
+
+    public boolean isNotAttendanceDay(final LocalDate localDate) {
+        return localDate.getDayOfWeek() == DayOfWeek.SUNDAY || localDate.getDayOfWeek() == DayOfWeek.SATURDAY || localDate.equals(LocalDate.of(2024, 12, 25));
+    }
```
여기도 테스트가 없어요... ㅠㅠ

verus-j commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,123 @@
+        final LocalDate localDate = LocalDate.of(localDateTime.getYear(), localDateTime.getMonthValue(),
+                localDateTime.getDayOfMonth());
+        if (localDateTime.getDayOfWeek() == DayOfWeek.SATURDAY || localDateTime.getDayOfWeek() == DayOfWeek.SUNDAY
+                || localDate.equals(christmas)) {
```
확실치는 않는데 크리스마스인 경우에 출석을 추가하지 않는 테스트가 없는 것 같네요

verus-j commented on src/main/java/domain/Operation.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.Arrays;
+import java.util.Objects;
+
+public enum Operation {
```
Operation에 대한 테스트가 없네요 ㅠㅠ

verus-j commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,180 @@
+
+    private void retryUntilOperationQuit() {
+        while (selectOperation() != Operation.QUIT) {
+        }
```
어떤 반복 작업을 하는지 블럭에 적힌 코드로는 확인할 수 없는 while문 이네요.

```
while (조건) {
    반복할 작업
}
```

보통 while문을 위와 같이 작성할 텐데, 현재는 조건을 확인하는 코드에 반복할 작업도 들어가 있는 것 같아요ㅠㅠ
어떻게 생각하시나요?

phk1128 replied:
저도 아쉬운 부분이라고 생각해요!
그래서 재귀를 사용하는 방법으로 리팩토링 했어요!

verus-j commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,180 @@
+            outputView.printNotAttendanceDay();
+            return;
+        }
+        final String crewName = LoopTemplate.tryCatchLoop(this::inputCrewName, outputView);
```
LoopTemplate을 재사용해서 예외 발생시 입력을 재시도할 수 있도록 하셨군요. 잘하셨습니다!
하지만 코드를 계속 보다보니 LoopTemplate이 어떤 용도인지 한눈에 알겠지만, inputCrewName에서는 입력값만 받고 있고, attendance에서는 입력 후 모델의 변경을 하고 있어 코드 파악이 어려운 것 같습니다 ㅠㅠ
예를 들어 LoopTemplate으로 입력값만 받는 작업을 반복하고 있다면 도메인 로직의 시작점을 파악하는데 더 도움이 될 것 같습니다.

verus-j replied:
이 부분은 어떻게 진행하시는걸까요??? 이모지만 있고 변경 사항이 없네요 ㅠㅠ
혹시 따로 윌슨이 생각하는 것이 있는걸까요?

phk1128 replied:
앗.. 변경사항이 푸시가 안되었나봐요! 수정 내용 푸시했어요!

verus-j commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,180 @@
+            final AttendanceResponse attendanceResponse = convertAttendanceToResponse(attendance);
+            outputView.printCrewAttendances(List.of(attendanceResponse));
+            return;
+        }
```
여기는 코드를 유심히 살펴보니 추가라는 동작을 수행하는데, 아래의 순서대로 동작하네요.

1. 출석 여부 확인
2. 뷰에 출석 시간 입력 및 출력
3. 출석 처리
4. 뷰에 결과 출력

느낌상 1번과 2번의 위치를 바꿔서 아래와 같이 도메인 로직을 한곳으로 모을 수 있을 것 같은데 어떻게 생각하시나요?
출석 여부 확인을 입력 후로 미뤄도 괜찮을 것 같아 제안드립니다.

1. 뷰 출력
2. 도메인 로직 (출석 여부 확인 및 출석 처리)
3. 뷰 출력

phk1128 replied:
출석 시간을 받기 전, 이미 출석을 완료한 경우 사용자에게 수정을 안내하기 위해 출석 여부를 미리 확인했어요
아래와 같이 진행이 되는데, 이미 출석을 했어도 시간을 입력 받는게 더 나은 로직일까요 ??
![image](https://github.com/user-attachments/assets/5fdf1d3a-edd1-4d34-88b0-73f74fef9840)

verus-j replied:
https://github.com/woowacourse/java-attendance/pull/33/files/739e498bde9c857c753e63e5aa4a9ad159bc856d#r1964846095

다시 살펴보니 위의 링크와 함께 봐야하는 코드였네요! 제가 구조 수정 제안을 드린 이유는 사용자의 편의성을 위해 코드를 복잡하게 작성해야하는가 입니다. 출석이 완료된 크루는 `이미 출석이 완료됐습니다.`라는 문구로 끝내면 안되는걸까요? 수정을 위한 기능은 이미 2번이 있지 않은가요?

물론 사용자의 편의성을 최대한 고려해주면 좋습니다. 하지만, 이를 위해 코드가 복잡해진다면 고민해야하는 부분입니다. 사용자에게 주는 가치와 코드의 복잡도를 비교 고민해 선택해야하는 문제지요.

제가 코드가 복잡하다고 느낀 이유는 제일 처음 피드백 드린 것과 같이 `출석 확인`이라는 동작 하나에 `모델 코드(1) - 뷰 코드(1) - 모델 코드(2) - 뷰 코드(2)`로 이어지기 때문에 모델 코드를 읽는 순간 문맥 전환이 발생합니다. 그리고 2번 째 모델 코드를 파악하기 위해서는 제일 앞에 있던 첫 번째 모델 코드의 내용도 기억하고 있어야합니다. 어쩌면 `출석 확인`의 기능이 변경될 때 이곳 저곳을 살펴봐야할 수도 있겠군요.

여기서 사용자 편의성을 약간 감수하면 `뷰 코드(1) - 모델 코드 - 뷰 코드(2)`와 같은 구조가 됩니다. 모델 코드를 한 곳에 모을 수 있어 이전보다 모델 코드를 파악하기 쉽다고 생각합니다. 추가로 모델 코드를 한 쪽에 모았으니 경우에 따라 컨트롤러에서 모델 코드를 사용하는 로직이 메서드 호출 하나로 해결할 수도 있습니다.

위의 점들을 고려하셔서 판단해보셨으면 합니다. 이정도 복잡도는 사용자의 편의성을 생각할 때 감안할만하다면 이대로 진행하면 됩니다. 하지만, 사용자 편의성을 위해 이정도의 복잡도를 증가시키는 방향이 옳지 않다면 고민해보시길 바랍니다.

phk1128 replied:
좋은 답변 감사해요!

초기화면에서 출석 수정 선택, 닉네임 입력 이 2스텝을 수정 선택(Y/N) 1스텝으로 줄이고자 이렇게 짰는데,

말씀을 들어보니 사용자의 편의성이 올라가는것에 비해 코드 복잡도가 너무 올라갔네요!

말씀하신대로 뷰 - 모델 - 뷰 구조로 리팩토링하는게 맞는 방향 같습니다!

아 그리고 혹시 이러한 상황과 관련해서 추천해주실만한 책이나 레퍼런스가 있을까요 ??

verus-j replied:
저 내용만 다룬 책이나 레퍼런스는 바로 떠오르지가 않네요 ㅠㅠ 그래도 결국에는 서로 다른 관심사는 분리하고, 같은 관심사끼리 한 곳에 모으는걸 집중하다보면 이러한 방향으로 흘러가는 것 같습니다!
조금 상황은 다르지만 `단위 테스트`라는 책의 아마도 7장에 이런 식으로 관심사를 분리한 후 한 곳에 모아 테스트 코드와 프로덕션 코드를 작성하는 방법을 설명하고 있을겁니다. 지금 당장 단위 테스트 책을 읽는 건 비추드리구요. 적어도 레벨 2를 다 겪고 난 뒤 읽는걸 추천드립니다. 우선은 `관심사 분리` 정도의 키워드로 인터넷에 검색해보는걸 추천드립니다!

verus-j commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,180 @@
+            outputView.printCrewAttendances(List.of(attendanceResponse));
+            return;
+        }
+        updateAttendanceForDuplicateAttendance(crewName);
```
크루의 이름이 중복인 케이스는 별도로 생각하신건가요?? 도메인에 대해 페어와 깊이 고민해보셨네요 ㅎㅎ

verus-j commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,180 @@
+        final LocalTime targetTime = LoopTemplate.tryCatchLoop(this::inputUpdateTime, outputView);
+        final Attendance beforeAttendance = attendanceSystem.findAttendanceByDate(crewName, dayOfMonth);
+        final Attendance afterAttendance = attendanceSystem.updateAttendanceByCrewNameAndDay(targetTime, crewName,
+                dayOfMonth);
```
출석 수정을 위해 attendanceSystem을 연속으로 두번 호출해야하네요! 2개의 메서드를 합쳐서 인터페이스의 크기를 줄여보는건 어떨까요?

phk1128 replied:
이전 출석 기록, 수정 후 출석 기록을 한번에 가져오는 메서드로 수정하려면 로직을 많이 갈아 엎어야 해요 ㅠ
이 부분은 좋은 방법이 떠오르는대로 리팩토링 할게요!

verus-j replied:
깔끔하게 정리해주셨네요~

verus-j commented on src/main/java/domain/AttendanceSystem.java:
```diff
@@ -0,0 +1,110 @@
+import java.util.List;
+import java.util.Objects;
+
+public class AttendanceSystem {
```
처음 AttendanceSystem을 봤을 때 다양한 기능을 넣은 슈퍼 클래스 같았는데 내부를 살펴보니 일급 컬렉션인거 같네요!
AttendanceSystem보다는 다른 좋은 이름을 쓰는게 좋을 것 같아요!

verus-j commented on .gitignore:
```diff
@@ -30,3 +30,5 @@ out/
 
 ### VS Code ###
 .vscode/
+
```
> 1번 질문 현재 서비스 계층이 없어서 dto 변환을 컨트롤러에서 담당하게 되었는데요. 이 경우 dto 변환을 위한 매퍼 클래스로 리팩토링하는게 좋은 방법일까요??

dto 변환을 어디둘지 고민하기 전에 DTO가 필요한 이유가 있을까요?? 뷰에서 바로 모델을 사용하면 안되는걸까요?

phk1128 replied:
MVC 원칙에 따라 뷰와 도메인 사이의 결합도를 낮추기 위해 dto를 사용했어요
뷰에 도메인이 노출되면 도메인 변경 시 뷰도 수정 될 위험이 있으니까요!

verus-j replied:
그럼 dto와 매퍼 클래스는 어디에 속하나요? 모델인가요? 뷰인가요? 
도메인이 변경될 때 dto 또는 매퍼의 변경은 괜찮은건가요??

https://developer.mozilla.org/ko/docs/Glossary/MVC

> 뷰는 앱의 데이터를 보여주는 방식을 정의합니다.
쇼핑 리스트 앱에서, 뷰는 항목이 사용자에게 보여지는 방식을 정의하며, 표시할 데이터를 모델로부터 받습니다.

mdn 문서에 정리된 MVC를 읽어보면 뷰에 대해 위와 같은 설명이 있습니다. DTO 변환 얘기지만 MVC 얘기를 한 이유는 굳이 뷰와 모델을 격리시키기 위해 DTO가 필요한건 아니기 때문입니다. 윌슨이 변환 로직을 어디둘 지 고민하고 계시지만, 변환 로직을 제거하고 뷰에서 모델의 정보를 가져오는 방법도 있습니다. 

그래도 윌슨의 원래 질문에 대해 답변을 드리면 dto를 현재 사용하는 변환 로직이 간단한 것 같습니다. 굳이 매퍼 클래스로 분리하지 않아도 될 것 같아 보입니다.

phk1128 replied:
> 그럼 dto와 매퍼 클래스는 어디에 속하나요? 모델인가요? 뷰인가요?
> 도메인이 변경될 때 dto 또는 매퍼의 변경은 괜찮은건가요??

dto와 매퍼 클래스는 mvc와 분리된 독립된 계층으로 보는게 맞다고 생각해요!
그리고 도메인이 변경될 때 dto와 매퍼가 수정되는건 자연스러운 상황아닐까요 ??
그래서 변경의 전파가 dto와 매퍼가 아닌 다른 계층으로 퍼지면 안되게끔 설계하는게 가장 중요한것 같아요
베루스 생각은 어떠신가요??


> mdn 문서에 정리된 MVC를 읽어보면 뷰에 대해 위와 같은 설명이 있습니다. DTO 변환 얘기지만 MVC 얘기를 한 이유는 굳이 뷰와 모델을 격리시키기 위해 DTO가 필요한건 아니기 때문입니다. 윌슨이 변환 로직을 어디둘 지 고민하고 계시지만, 변환 로직을 제거하고 뷰에서 모델의 정보를 가져오는 방법도 있습니다.

뷰에서 모델을 참조하는것도 틀린 방법은 아니지만, 되도록 dto를 둬서 모델과 뷰는 분리를 해야한다고 생각해요.
뷰는 단위테스트도 힘들고 변경이 일어나면 사용자한테도 전파될 수 있으니까요. 반면에 dto의 경우 단위테스트가 쉬워서 변경이 일어나면 빠르게 검증 가능하고, 이를 통해 변경 전파가 다른 계층으로 퍼지는것도 막기 수월하다고 생각해요
결과적으로 되도록 dto를 두는게 변경에 대해 안전하지 않을까 생각해요!

verus-j replied:
> 도메인이 변경될 때 dto와 매퍼가 수정되는건 자연스러운 상황아닐까요 ??
> 그래서 변경의 전파가 dto와 매퍼가 아닌 다른 계층으로 퍼지면 안되게끔 설계하는게 가장 중요한것 같아요

도메인이 변경될 때 DTO, 매퍼, 뷰의 변경은 자연스러운 상황이 맞습니다! 저는 실은 도메인이 변경될 때 어느 한곳은 변경되어야 한다를 얘기해보고 싶었습니다.(어디에 속한 계층인지 물어본건 윌슨의 생각이 궁금해서였습니다!) 뷰에 어떻게 정보를 전달하냐에 따라 DTO, 매퍼, 뷰를 변경하게 될 텐데 현재 상황에서는 어떤 방식이 특별히 좋다고 판단할 근거가 없다고 생각했습니다.

만약에 모델과 뷰에서 사용할 데이터의 구조가 많이 다르다면, DTO를 둬서 변환 과정을 따로 관리할 수 있습니다. 하지만, 현재 구조에서는 DTO는 단순 모델과 동일한 데이터 구조를 가진 정보를 전달하고 있습니다. 그렇기에 모델을 변경했을 때 모델을 바로 사용하는 뷰를 변경하나 DTO를 변경하나 특별히 어떤 것이 좋다고 판단할 수 없다고 느꼈습니다.

> 1번 질문 현재 서비스 계층이 없어서 dto 변환을 컨트롤러에서 담당하게 되었는데요. 이 경우 dto 변환을 위한 매퍼 클래스로 리팩토링하는게 좋은 방법일까요??

그래서 실은 위의 질문에 대한 답변으로 매핑을 위한 로직에 대한 적절한 위치를 찾지 못해 고민하고 있다면 그냥 DTO나 매퍼를 사용하지 않는게 방법일 수도 있다를 얘기해보고 싶었습니다.

verus-j commented on .gitignore:
```diff
@@ -30,3 +30,5 @@ out/
 ### VS Code ###
 .vscode/
+
+bin/
```
> 2번 질문 애플리케이션 종료 후 resources 패키지 존재하는 csv파일을 원상태로 복구하는게 좋을까요? 아니면 변경 내용들을 csv파일에 저장하는게 좋을까요 ??

이 부분은 현재 상황에서 판단하기 어려운 것 같습니다. 이유는 애플리케이션이 동작중에 변경한 데이터를 유지해야하는 서비스인지 아닌지를 판단하기 어렵기 때문인 것 같습니다. 출석 시스템을 생각해보면 데이터를 유지해야 하지만, 간단한 우테코 미션을 생각해보면 불필요한 기능이라고 생각됩니다. 

우선 이번 미션에서는 애플리케이션이 변경한 데이터를 유지(영속성)하는 건 고려하지 않고 진행하시죠! 추후 다른 미션들(아마도 레벨2)에서 영속성을 고려한 미션들이 나옵니다. 그 때 이번 미션에서 파일을 저장하는게 좋았을 지 다시 판단해보면 좋을 것 같습니다.

phk1128 replied:
명쾌한 답변 감사합니다!

verus-j commented on README.md:
```diff
@@ -1,3 +1,28 @@
 # java-attendance
 
 출석 미션 저장소
+
```
> 3번 질문 Attendance 클래스 생성자에서 매개변수로 yyyy-MM-dd HH:mm 형태의 문자열을 받아 LocalDateTime으로 파싱하는 작업을 하고 있어요. 여기서 외부에서 날짜를 파싱해서 LocalDateTime을 Attendance 생성자로 넘겨주는게 좋을것 같은데, 베루스의 생각은 어떠신가요??

메서드의 매개변수를 고민할 때는 어떤 방법이 쉽게 사용하고 헷갈리지 않을지를 고민하는 것 같습니다. 문자열 yyyy-MM-dd HH:mm 형태로 받게 되면 메서드(생성자)를 호출하는데 실수를 많이 할 것 같습니다. 그래서 LocalDateTime을 사용하는게 Attendance 생성자를 사용하기 더욱 쉬울 것 같습니다.

verus-j reviewed (changes requested):
안녕하세요 윌슨
모든 커밋을 다 확인할 수 없었지만 초기의 커밋을 보면 TDD 사이클을 지키려고 하신 모습이 보이네요!
미션 수행하시느라 고생많으셨습니다.
몇가지 피드백과 질문에 대한 답 남겨뒀습니다!
확인부탁드립니다~

phk1128 committed:
test(AnswerCommandTest): 응답 커멘드 성공 및 실패 테스트

phk1128 committed:
test(AttendanceSystemTest): 크루 존재 확인 및 유효한 날짜 성공 테스트

phk1128 committed:
test(AttendanceSystemTest): 등교일 확인 테스트

phk1128 committed:
refactor: 주말과 공휴일은 출석 확인이 안되게 수정 및 테스트

phk1128 committed:
test(Operation): Operation 객체 성공 및 실패 테스트

phk1128 committed:
refactor: Attendance 생성자 시그니처 변경에 따른 리팩토링

- 문자열을 LocalDateTime으로 파싱하는 DateTimeParser 객체 구현
- Attendance 생성자 매개변수를 String 타입에서 LocalDateTime으로 변경

phk1128 committed:
feat(HoliDay, HoliDayTest): 주말 및 공휴일를 정의한 enum 구현

테스트 진행 내용
- 주말 또는 공휴일을 확인한다

phk1128 committed:
feat(ClassTime, ClassTimeTest): 강의 시간을 정의한 enum 구현

테스트 진행 내용
- 요일에 맞는 강의 시간을 가져온다
- 주말과 공휴일에 강의 시간을 조회하여 예외가 발생한다

phk1128 committed:
refactor: beforeAttendance, afterAttendance를 하나의 메서드로 처리

리팩토링 내용
- 주말 및 공휴일에 출석 확인 및 수정을 하면 예외가 발생하도록 수정
- 출석 수정 시 AttendanceSystem의 메서드가 2번 호출되는 부분을 1번 호출하게끔 수정
- 중복 메서드 제거

phk1128 committed:
refactor(AttendanceController): LoopTemplate를 사용하는 출석 시간 입력과 출석 비즈니스로직 메서드를 분리

phk1128 commented:
안녕하세요 베루스!

리뷰해주셔서 감사해요 👏

피드백 주신 내용을 기반으로 1차 리팩토링 진행해봤어요

- [x] 성공 또는 실패 테스트 누락된 부분 작성
- [x] 출석 수정 시 AttendanceBook(= AttendanceSystem)의 메서드 1개만 호출 하도록 수정
- [x] 주말 및 공휴일에 출석 확인 및 수정을 하면 예외가 발생하도록 수정
- [x] attendance() 메서드 시간 입력과 출석 로직 LoopTemplate 분리
- [x] 주말 및 공휴일을 정의한 enum 추가
- [x] 강의 시간을 정의한 enum 추가 

그럼 리뷰 부탁드려요!

verus-j commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    private void retryUntilOperationQuit() {
+        if (selectOperation() != Operation.QUIT) {
```
selectOperation은 조회 메서드인가요? 명령 메서드인가요? selectOperation은 operation을 조회하는 것 같은데, 내부에는 객체의 상태 변경(명령)도 있네요... ㅠㅠ 재귀로 바꿨지만, 이전의 while문을 사용할 때와 크게 달라진 점이 없는 것 같습니다.

클린 코드 3장 함수(p56)을 살펴보면 아래의 내용이 있습니다.

> **명령과 조회를 분리하라!**
> 함수는 뭔가를 수행하거나 뭔가에 답하거나 둘 중 하나만 해야 한다. 둘 다 하면 안 된다. 객체 상태를 변경하거나 아니면 객체 정보를 반환하거나 둘 중 하나다. 둘 다 하면 혼란을 초래한다.

명령과 조회를 분리하는게 절대적인 법칙은 아니지만, 지금은 개선할 필요가 보입니다. 참고하셔서 어떻게 조회와 명령을 분리할 지 고민해보시면 좋을 것 같습니다!

verus-j commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,73 @@
+    public Attendance(final LocalDateTime dateTime, final boolean isEmpty) {
+        this.dateTime = dateTime;
+        this.isEmpty = isEmpty;
+    }
```
주석들은 어떤 의미인가요? 이미 생성자가 주고 있는 의미 아닌가요?

phk1128 replied:
지우는걸 깜빡했네요 😅

verus-j commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,123 @@
+        final Attendance attendance = new Attendance(DateTimeParser.parseToLocalDate(attendanceTime));
+        attendances.add(attendance);
+        return attendance;
+    }
```
AttendanceBook에서 LocalDateTime을 문자열로 변환해 addAttendance로 전달하고, 여기서 다시 LocalDate로 변환하고 있네요. LocalDateTime을 그대로 받아서 LocalDate로 변환(`localDateTime.toLocalDate()`)하면 안되는건가요?

phk1128 replied:
그렇네요 ㅎㅎ 코드를 정말 꼼꼼히 보셨나봐요~
감사합니다 이 부분도 리팩토링에 반영할게요!

verus-j commented on src/main/java/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,127 @@
+        validateAttendanceDate(localDateTime);
+        final Crew crew = findCrewByName(name);
+        return crew.addAttendance(localDateTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")));
+    }
```
이번에 드린 다른 피드백과도 연관되는 부분인데요. attendance는 명령인가요? 조회인가요?

phk1128 replied:
`검증 -> 크루 찾기 -> 출석 확인` 흐름의 로직이에요
조회와 명령이 함께 있어서 단일 책임 원칙에 위배되는 상황이라 메서드를 쪼갤 수 있지만,
크루를 찾고 해당 크루를 출석하는 흐름이 간단하고 자연스럽기 때문에 메서드를 쪼개는게 오히려 복잡해질 것 같아요
베루스의 생각은 어떠신가요 ??!

verus-j replied:
네네 이정도면 충분합니다! 좋은 사유를 가지고 잘 판단하신 것 같습니다!

verus-j commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,64 @@
+    public Attendance(final LocalDateTime localDateTime) {
+        this.dateTime = localDateTime;
+        isEmpty = false;
+    }
```
생성자 체이닝을 해보는건 어떨까요?

verus-j commented on src/main/java/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,127 @@
+                                                       final int dayOfMonth) {
+        final LocalDate localDate = LocalDate.of(2024, 12, dayOfMonth);
+        validateAttendanceDate(localDate);
+        final Attendance beforeAttendance = findAttendanceByDate(crewName, dayOfMonth);
```
dayOfMonth를 findAttendanceByDate에 넣고 있는데, 내부에 들어가보니 LocalDate.of(2024, 12, dayOfMonth)를 다시 생성하고 있네요. localDate를 그대로 전달하면 안될까요?

verus-j commented on src/main/java/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,127 @@
+        final LocalDate localDate = LocalDate.of(2024, 12, dayOfMonth);
+        validateAttendanceDate(localDate);
+        final Attendance beforeAttendance = findAttendanceByDate(crewName, dayOfMonth);
+        final Crew crew = findCrewByName(crewName);
+        crew.updateAttendanceByDateAndTime(targetTime, localDate);
```
위에는 수정전 출석 정보를 조회하기 위해 findAttendanceByDate라는 메서드로 추상화했는데, 여기는 수정을 위해 크루를 찾고, 업데이트하는 findAttendanceByDate보다 로우한 코드인 것 같네요. 추상화 수준을 동일하게 맞춰주는 건 어떨까요?

verus-j commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,123 @@
+
+    public void updateAttendanceByDateTime(final String attendanceTime) {
+        final Attendance updatedAttendance = Attendance.of(attendanceTime);
+        attendances.remove(updatedAttendance);
+        attendances.add(updatedAttendance);
```
동일한 객체를 제거하고 더하나요? 이렇게 되면 아무것도 안한것과 동일한거 아닌가요?

phk1128 replied:
Attendance 객체에 equals를 LocalDate가 같으면 동등하게 보도록 오버라이드 했어요
구체적으로는 이전 출석 기록 지우고 시간이 갱신된 업데이트 된 Attendance 객체를 추가하는거였어요

근데 베루스님이 짚어주셔서 다시 보니, 코드를 읽는 사람이 충분히 오해할만한 소지가 있는 로직이네요!
해당 부분은 좀 더 가독성있게 리팩토링해볼게요!

verus-j reviewed (changes requested):
안녕하세요 윌슨~
코드 반영해주신거 확인했고 몇가지 피드백 더 남겼습니다.
확인부탁드립니다!

phk1128 committed:
refactor(AttendanceController): selectOperation() 메서드 제거 및 조회 변경 로직 분리

phk1128 committed:
refactor(Crew): 출석 수정 시 이전 출석 기록이 있는 경우에만 remove 하도록 수정

phk1128 committed:
refactor(Crew): 스트림을 이용하여 이전 출석 기록 삭제하도록 수정

phk1128 committed:
refactor: addAttendance()의 매개변수 시그니처를 String에서 LocalDateTime 타입으로 변경

phk1128 committed:
refactor(AttendanceBook): 크루 출석 수정 메서드 분리

phk1128 committed:
refactor(AttendanceBook): findAttendanceByDate() 매개변수 시그니처 String에서 LocalDate로 변경에 따른 리팩토링

phk1128 committed:
refactor(Attendance): 생성자 체이닝 사용

phk1128 committed:
refactor: 이미 출석한 경우 안내문만 출력 후 초기화면으로 돌아가도록 수정

phk1128 reviewed (commented):
안녕하세요 베루스!

2차 리뷰도 정말 꼼꼼하게 해주셔서 감사합니다!

말씀해주신 부분 기반으로 리팩토링 진행했으니 확인 부탁드려요!

verus-j commented on src/main/java/domain/ClassTime.java:
```diff
@@ -0,0 +1,36 @@
+
+    private final DayOfWeek dayOfWeek;
+    private final LocalTime startTime;
+    private final LocalTime endTime;
```
여기 endTime은 사용하지 않고 있네요!

verus-j commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,41 @@
+        final ClassTime classTime = ClassTime.findClassTimeByDateTime(dateTime.toLocalDate());
+        final LocalDateTime timeBoundary = dateTime
+                .withHour(classTime.getStartTime().getHour())
+                .withMinute(classTime.getStartTime().getMinute());
```
여기는 ClassTime으로 옮길 수 있는 메서드 같습니다!
어쩌면 ClassTime을 구하지 않고 바로 정적 메서드로 timeBoundary를 구할 수도 있겠네요.

verus-j commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,41 @@
+                .sorted((o1, o2) -> o2.boundary - o1.boundary)
+                .filter(status -> dateTime.isAfter(timeBoundary.plusMinutes(status.boundary)))
+                .findFirst()
+                .orElse(ATTENDANCE);
```
위의 로직을 ClassTime으로 위임하고 나면 마지막 남은 Stream 문이 추상화 수준이 안 맞을 것 같습니다. 위의 작업을 하고 난뒤 여기도 추출을 고려해보시죠!
```java
    public static AttendanceStatus of(final LocalDateTime dateTime) {
        final ClassTime classTime = ClassTime.findClassTimeByDateTime(dateTime.toLocalDate());
        final LocalDateTime timeBoundary = classTime.timeBoundary()
        return Arrays.stream(values())
                .sorted((o1, o2) -> o2.boundary - o1.boundary)
                .filter(status -> dateTime.isAfter(timeBoundary.plusMinutes(status.boundary)))
                .findFirst()
                .orElse(ATTENDANCE);
    }
```

verus-j commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.List;
+
+public enum AttendanceStatus {
+    ATTENDANCE("출석", 0),
+    LATE("지각", 5),
+    ABSENCE("결석", 30);
```
"출석", "지각", "결석" 문자열은 뷰를 위한 정보가 아닐까요?? 만약 사용자의 언어가 바뀌어 영어로 수정해야한다면 어떻게 될까요?

phk1128 replied:
각 문자열은 표현을 위한 데이터이긴 하지만 ATTENDANCE, LATE, ABSENCE 각 단어의 뜻이라서 응집도 측면에서 좋을것 같아서 enum의 요소로 두었어요
그리고 단순히 표현을 위한거라 수정이 일어나도 비즈니스 로직 또는 다른 계층에 영향을 주는게 한정적이라고 생각했어요. 
하지만 말씀하신대로 사용자의 언어가 바뀐다면 매번 해당 언어에 맞게끔 수정해야하는 단점이 존재하네요.

반면에 뷰에서 관리 할 경우 `ResourceBundle`을 통해 국제화를 이용할 수 있어서 사용자의 언어에 따라 매번 수정할 필요가 없는 장점이 존재한다고 하는데요
그렇다면 결국에는 국제화 장점을 위해 OutputView에 있는 모든 메세지를 `ResourceBundle`를 이용하는걸로 수정해야 하는걸까요 ??

verus-j replied:
네네 현업에서는 다국어 적용할 때 얘기하신 방법으로 관리합니다! 
실은 저기서 굳이 다국어 얘기를 꺼낸건 해당 정보들이 뷰를 위한 정보임을 정당화(?)하기 위해서 가져온 사례입니다. 당연히 우테코 미션에서는 필요없는 기능이죠! 

> 그리고 단순히 표현을 위한거라 수정이 일어나도 비즈니스 로직 또는 다른 계층에 영향을 주는게 한정적이라고 생각했어요.
> 하지만 말씀하신대로 사용자의 언어가 바뀐다면 매번 해당 언어에 맞게끔 수정해야하는 단점이 존재하네요.

현재 구조에서는 다른 계층에 영향을 주는게 한정적이라고 판단하셨고, 앞으로 어떤 문제가 발생할 수 있는지 인지했으니 충분한 것 같습니다. (미션은 여기서 끝이지만) 추후 상황이 발생하면 생각하신 방법들로 대처해볼 수 있을 것 같습니다.

phk1128 replied:
명쾌한 답변 감사합니다!

verus-j commented on src/main/java/domain/ExpulsionStatus.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.Arrays;
+
+public enum ExpulsionStatus {
+    NORMAL("정상", 0),
+    ADVANCE("경고", 2),
+    INTERVIEW("면담", 3),
+    EXPULSTION("제적", 6);
```
여기도 각각의 문자열이 뷰를 위한 정보 아닌가요?? 개념은 동일하지만 명칭만 변경되면 어떻게 될까요? 예를 들어 "경고"라는 문구가 "주의"로 바뀌고, "면담"이 "상담"과 같이 바뀐다면요. 이러한 변경은 도메인에 대한 변경은 아니고 뷰만 변경해야할 것 같습니다.

phk1128 replied:
관련해서 위에 제 의견 남겨두었습니다!

verus-j commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,169 @@
+    }
+
+
+    private void responseCrewAttendanceHistory() {
+        final String crewName = LoopTemplate.tryCatchLoop(this::inputCrewName, outputView);
+        final Crew crew = attendanceBook.findCrewByName(crewName);
+        outputView.printAttendanceHistoryTitle(crewName);
+        final List<AttendanceResponse> attendanceResponses = convertAttendancesToResponses(crew.getAttendances());
+        outputView.printCrewAttendances(attendanceResponses);
+        final Map<AttendanceStatus, Integer> attendanceStatistics = crew.calculateAttendanceStatistics();
+        outputView.printAttendancesStatistics(attendanceStatistics);
+        final ExpulsionStatus expulsionStatus = crew.calculateExpulsionStatus();
+        outputView.printCrewExpulsionStatus(expulsionStatus);
+
+    }
```
코드를 보니 호출 순서를 살짝 바꾸면 outputView의 인터페이스를 줄일 수 있을 것 같아 보입니다. 가능할지 확인 부탁드려요~

```suggestion
    private void responseCrewAttendanceHistory() {
        final String crewName = LoopTemplate.tryCatchLoop(this::inputCrewName, outputView);
        final Crew crew = attendanceBook.findCrewByName(crewName);
        final List<Attendance> attendances = crew.getAttendances();

        // crewName과 dto를 같이 넘겨서 outputView의 인터페이스를 줄일 수 있을 것 같습니다.
        outputView.printAttendanceHistoryTitle(crewName);
        outputView.printCrewAttendances(convertAttendancesToResponses(attendances));

        final Map<AttendanceStatus, Integer> attendanceStatistics = crew.calculateAttendanceStatistics();
        final ExpulsionStatus expulsionStatus = crew.calculateExpulsionStatus();

        // 여기 역시 attendanceStatistics와 expulsionStatus를 같이 넘겨 outputView의 인터페이스를 줄일 수 있을 것 같습니다.
        outputView.printAttendancesStatistics(attendanceStatistics);
        outputView.printCrewExpulsionStatus(expulsionStatus);

    }
```

verus-j reviewed (changes requested):
안녕하세요 윌슨!
코드가 정말 많이 깔끔해졌네요~
1단계도 거의 끝나가는 것 같습니다.
소소한 피드백 남겨드렸습니다. 확인부탁드립니다~

phk1128 committed:
refactor: OutputView 인터페이스 간소화에 따른 리팩토링

phk1128 committed:
refactor: AttendanceStatus의 of 메서드 간소화

phk1128 reviewed (commented):
안녕하세요 베루스!

3차도 리뷰도 꼼꼼하게 해주셔서 감사합니다 ㅎㅎ

피드백 주신 내용 바탕으로 리팩토링 진행하였고, 관련해서 제 의견도 적어보았어요

혹시 이외에 추가적으로 피드백해주실 내용 있으시면 말씀 부탁드리고 해당 내용들은 step2에 반영하도록 할게요!

그럼 step2도 잘부탁드립니다 :)

verus-j reviewed (approved):
안녕하세요 윌슨!
1단계 진행하시느라 고생 많으셨습니다~
피드백을 많이 드렸는데, DM 내용을 보니 도움이 되신 것 같아 기쁘네요 ㅎㅎ
여기서 이만 머지하고, 2단계에서 뵙겠습니다!
